### PR TITLE
Menu when encountering an army

### PR DESCRIPTION
**Pull request by JoeFwd**
Implementation of https://github.com/Tyler-IN/MnB2-Bannerlord-CommunityPatch/issues/120#issuecomment-613501060

A menu will pop up when encoutering a **neutral** army enabling the player to : 
* talk to the army's leader
* talk to the other members
* leave

A menu will pop up when encoutering an **enemy** army enabling the player to : 
* talk to the army's leader
* attack
* leave

### CODE REVIEW

**Comment by Tyler-IN** [comment]
I approved the code, but does this need to be an option-toggled feature or is it TW's intention?

File: src/CommunityPatch/Patches/MenuWhenEncountingAnArmyPatch.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    
+    
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static bool Prefix(PlayerEncounter __instance) {
+      if (EncounteredPartyField == null || MapEventStateField == null || StateHandledField == null || DefenderPartyField == null || MeetingDoneField == null) {
+        CommunityPatchSubModule.Error($"{typeof(MenuWhenEncounteringAnArmyPatch).Name}: Could not locate all of necessary private fields for patching." + Environment.NewLine);
+        return true;
+      }
+      
+      PartyBase party1 = (PartyBase) EncounteredPartyField.GetValue(__instance);
```
Comment (tynakuh) [inline src/CommunityPatch/Patches/MenuWhenEncountingAnArmyPatch.cs:75]: party1 vs party2 aren't very descriptive and can make it a bit harder to follow imo.

Comment (JoeFwd) [reply src/CommunityPatch/Patches/MenuWhenEncountingAnArmyPatch.cs:75]: Tbh I left the vanilla implementation names.

Comment (Tyler-IN) [reply src/CommunityPatch/Patches/MenuWhenEncountingAnArmyPatch.cs:75]: We can refactor the naming to attacker/defender or whatever this after it gets merged to dev, or...

File: src/CommunityPatch/Patches/MenuWhenEncountingAnArmyPatch.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      if (party1.IsSettlement)
+      {
+        foreach (PartyBase party2 in (IEnumerable<PartyBase>) MobileParty.MainParty.MapEvent.DefenderSide.Parties)
+        {
+          if (!party2.IsSettlement)
+          {
+            party1 = party2;
```
Comment (Tyler-IN) [inline src/CommunityPatch/Patches/MenuWhenEncountingAnArmyPatch.cs:82]:
```suggestion
      PartyBase attacker = (PartyBase) EncounteredPartyField.GetValue(__instance);
      if (attacker.IsSettlement)
      {
        foreach (PartyBase otherParty in (IEnumerable<PartyBase>) MobileParty.MainParty.MapEvent.DefenderSide.Parties)
        {
          if (!otherParty.IsSettlement)
          {
            attacker = otherParty;
```

**Comment by Tyler-IN** [comment]
Holding this PR for now, this should probably be an opt-in feature, right?

**Comment by Tyler-IN** [comment]
Opt-in means it should be disabled by default, `DisableMenuWhenEncouteringAnArmy ` defaults to false right? So you should probably flip all the checks and rename it to `EnableMenuWhenEncouteringAnArmy` ... At current, it looks like it's opt-out instead of opt-in.

**Comment by JoeFwd** [comment]
> 
> 
> Opt-in means it should be disabled by default, `DisableMenuWhenEncouteringAnArmy ` defaults to false right? So you should probably flip all the checks and rename it to `EnableMenuWhenEncouteringAnArmy` ... At current, it looks like it's opt-out instead of opt-in.

Yes, you are right. I'll flip the opt !

**Pull request merged**